### ISSUE TITLE
[Tech spec] Include top-of-file comments for .rb, .js, and .jsx files

### ISSUE THREAD

**Issue by joeyyang** [department-of-veterans-affairs/caseflow#9619]
**tl;dr:** I'd like to propose we add top-of-file comments `.js(x)` and `.rb` files and method-level comments in `.js` and `.rb` files, consistent with [JSDoc](http://usejsdoc.org/about-getting-started.html) and [rdoc](https://github.com/ruby/rdoc) standards, respectively.

### Description
One thing I've struggled with on this project has been documentation in and about code. There is a lot of complex logic representing complex operational processes at BVA, and I've had to rely on the very friendly people on this project to guide me through the data models and control flow of the application, even for relatively simple tasks. 

To preserve institutional knowledge since we're past AMA day and many folks will be rolling off the project, I'd like to propose a coordinated effort to comment the codebase. 

I'd like to propose we add comments:
-  at the top of the file in:
   - all `.rb` files under `/app`
   - all `.js` and `.jsx` files under `/client/app`
- ~~at the top of any method over 1 line long in:~~
  - ~~all `.rb` files under `app`~~
  - ~~all `.js` files under `client/app` (include method comments in `.jsx` files only when necessary)~~

### Implementation proposal
As I'm rolling off the project soon, I will take a few days to add comments to all the code I've touched. I'd like to propose that other team members leaving the project take a few days to do this as well. Otherwise, teams could designate a part of a sprint to dedicate to documentation and other clean-up tasks.

### Documentation standards
I am less familiar with Ruby standards, but for JSDoc, add a sentence-length comment so the reader understands what each piece of code does at a glance. (I don't think we need to go to the level of type and parameter annotation, since we just removed Flow.)

For example, in `client/app/queue/selectors.js`: 
```javascript
/**
 * Redux selectors used to pull data from the Redux store for Caseflow Queue.
 * Selectors are used in React components to derive current state of a part of the application from the global state object.
 */

import { createSelector } from 'reselect';
import _ from 'lodash';
import {
  taskIsActive,
  taskIsOnHold
} from './utils';

import TASK_STATUSES from '../../constants/TASK_STATUSES.json';

/**
 * Get all tasks selected by user in Queue TaskTable UI. 
 * Determines which tasks to update upon user action.
 */
export const selectedTasksSelector = (state, userId) => {
  return _.map(
    state.queue.isTaskAssignedToUserSelected[userId] || {},
    (selected, id) => {
      if (!selected) {
        return;
      }

      return state.queue.tasks[id] || state.queue.amaTasks[id];
    }
  ).filter(Boolean);
};

const getTasks = (state) => state.queue.tasks;
const getAmaTasks = (state) => state.queue.amaTasks;
...
```

In `app/models/claim_review.rb`: 
```ruby
##
# A claim review is a short hand term to refer to either a supplemental claim or
# higher level review as defined in the Appeals Modernization Act of 2017

class ClaimReview < DecisionReview
  include HasBusinessLine

...

  # Serialize claim review data to be displayed in various places in the Caseflow UI. 
  def ui_hash
    super.merge(
      benefitType: benefit_type,
      payeeCode: payee_code,
      hasClearedEP: cleared_ep?
    )
  end
```

In `client/app/queue/components/HearingBadge.jsx`:
```jsx
/**
 * A badge that displays in TaskTables that indicates if a hearing has been held for a claim. 
 * Shows date of hearing, presiding judge, and hearing type.
 * Only shows up if the hearing has {disposition: 'held'}. 
 */
import PropTypes from 'prop-types';
import { css } from 'glamor';
import _ from 'lodash';
import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
      appeal,
      tasks
    } = this.props;

    let sectionBody = COPY.TASK_SNAPSHOT_NO_ACTIVE_LABEL;

    if (tasks.length) {
      sectionBody = <table {...tableStyling}>
```

**Comment by pkarman** [comment]
I'm +1 on the top-of-the-file comments to orient folks about the purpose of the file.

I'm +0 on the per-method comments. If the method name doesn't communicate it well enough, that seems like a bigger problem.

**Comment by monfresh** [comment]
Big picture comments are a great idea. The fact that the code is hard to read and understand is a code quality issue that is best solved by refactoring the code so that it is more readable and doesn't require comments that explain what the method does. Comments can be valuable when they explain the "why", but even that can usually be addressed by a well-named method. Here is a great example: https://thoughtbot.com/blog/code-that-says-why-it-does

Refactoring is not something that gets scheduled once in a while. The goal is to improve the code little by little every day. Something as simple as renaming a method, or breaking up complex compound conditionals into well-named methods can go a long way. 

If we all adopt the good habit of leaving code better than we found it, and refactoring as we go along, the code base will improve over time. One way to do that is for all of us to participate in the awesome [30-Day Code Quality Challenge](https://www.codequalitychallenge.com/).

Here are some more of my favorite resources on code quality and refactoring:

https://thoughtbot.com/upcase/clean-code
[Ben Orenstein - Refactoring from Good to Great](https://www.youtube.com/watch?v=DC-pQPq0acs)
https://www.refactoringrails.io/
[Sandi Metz - Polly Want A Message](https://www.deconstructconf.com/2018/sandi-metz-polly-want-a-message)
[Sarah Mei - Livable Code](https://www.youtube.com/watch?v=lI77oMKr5EY)
[Katrina Owen - Here Be Dragons](https://www.youtube.com/watch?v=3ZHN2btOjL8)
Choice quote from Katrina's talk:

> Every commit we make tips the balance ever so slightly in one direction or the other. An inch towards entropy. An inch towards increased order. Perfection is unattainable, and ultimately irrelevant. If most of the commits you make are shifting the balance towards increased order, then no matter how slight that shift is, you will have a codebase that constantly improves.

This whole [Twitter thread by Sarah Mei](https://twitter.com/sarahmei/status/1090860150712221698) is gold. Here are some quotes that stood out for me:

>The most common “presenting” pathology in the hoarded codebases I’ve seen - by far - is that developers don’t feel they have time and/or permission to refactor code. Frequently occurring alongside that pathology is another - that developers see “refactoring” as a completely separate activity from building features or fixing bugs. A key indicator of this pathology is seeing stories in the backlog like “refactor user class.”

> Just like its physical analog, a hoarded codebase only improves if you intensively work on changing those habits. This means deciding you will always do small, opportunistic refactorings when they appear to you in the course of fixing a bug or adding a feature.

> I’m not talking about taking three extra days on a 1-point story to totally rewrite the user class. I’m talking about noticing a method you’re working in is out of place, and moving it - even if you don’t have time to extract the rest of the concept from the 8000-line file.

> Just like when you’re dealing with its physical analog, your number one most important mantra when you want to improve a hoarded codebase is: Improvement Over Consistency. One book on the shelf and five in the pile is better than six books in the pile. Improvement Over Consistency.

**Comment by joeyyang** [comment]
Thanks @monfresh and @pkarman! **I will amend the above spec to only call for file-level comments.**

I agree that if methods are named correctly that we shouldn't need to comment the methods themselves, and within a culture of continuous improvement we shouldn't need refactoring sprints. In that spirit, I would encourage folks to rename confusing method names if they touch them.  

I am a fan of something like the 30-Day Code Quality Challenge @monfresh points to, especially at this time in the project lifecycle. I think it would be one of the best things we can do to ensure this project is successful well into the future.
